python-package/xgboost/sklearn.py

```diff
@@ -1100,12 +1100,14 @@ def get_num_boosting_rounds(self) -> int:
         return DEFAULT_N_ESTIMATORS if self.n_estimators is None else self.n_estimators
 
     def _get_type(self) -> str:
-        if not hasattr(self, "_estimator_type"):
-            raise TypeError(
-                "`_estimator_type` undefined.  "
-                "Please use appropriate mixin to define estimator type."
-            )
-        return self._estimator_type  # pylint: disable=no-member
+        if hasattr(self, "_estimator_type"):
+            return self._estimator_type  # pylint: disable=no-member
+        if hasattr(XGBModelBase, "__sklearn_tags__"):
+            return self.__sklearn_tags__().estimator_type
```

Comment:
For version changes like this you might want to leave a comment on what version of sklearn each branch supports (or branch on versioning specifically) so you know when you can drop the compatibility code.